app/acme/dnsprovider/cloudns.go

```diff
@@ -0,0 +1,289 @@
+		"domain-name": record.Domain,
+		"record-id":   strconv.Itoa(r.id),
+	}
+
+	b, err := p.doRequest(http.MethodPost, removeRecordURL, params)
+	if err != nil {
+		return fmt.Errorf("remove record failed: %s", err)
+	}
+
+	res := struct {
+		Status            string `json:"status"`
+		StatusDescription string `json:"statusDescription"`
+	}{}
+
+	if err := json.Unmarshal(b, &res); err != nil {
+		return fmt.Errorf("remove record failed: %s", err)
+	}
+
+	if res.Status != "Success" {
+		return fmt.Errorf("remove record failed: %s: %s", res.Status, res.StatusDescription)
+	}
+
+	recs := p.addedRecords[:0]
+	for _, rec := range p.addedRecords {
+		if rec.id != r.id {
+			recs = append(recs, rec)
+		}
+	}
+	p.addedRecords = recs
+
+	return nil
+}
+
+func (p *cloudnsProvider) WaitUntilPropagated(ctx context.Context, record Record) error {
+	timeout, interval := p.GetTimeout()
+	ticker := time.NewTicker(interval)
+	timer := time.NewTimer(timeout)
+
+	for {
+		select {
+		case <-ticker.C:
+			updated, err := p.isUpdated(record.Domain)
+			if err != nil {
+				return err
+			}
+			if updated {
+				return nil
+			}
+		case <-ctx.Done():
+			return fmt.Errorf("timeout waiting for DNS propagation")
+		case <-timer.C:
+			return fmt.Errorf("timeout waiting for records update")
+		}
+	}
+}
+
+func (p *cloudnsProvider) isUpdated(domain string) (bool, error) {
+	params := map[string]string{
+		"domain-name": domain,
+	}
+
+	b, err := p.doRequest("GET", updateStatusURL, params)
+	if err != nil {
+		return false, err
+	}
+
+	servers := []struct {
+		Server  string `json:"server"`
+		IP4     string `json:"ip4"`
+		IP6     string `json:"ip6"`
+		Updated bool   `json:"updated"`
+	}{}
+
+	if err := json.Unmarshal(b, &servers); err != nil {
+		return false, err
+	}
+
+	updated := 0
+	for _, server := range servers {
+		if server.Updated {
+			updated++
+		}
+	}
+	return updated == len(servers), nil
+}
+
+func (p *cloudnsProvider) doRequest(method, endpoint string, params map[string]string) (json.RawMessage, error) {
+	reqURL, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	q := reqURL.Query()
+
+	for k, v := range params {
+		q.Set(k, v)
+	}
+
+	// these should be set for all requests
+	q.Set("sub-auth-id", p.subAuthID)
+	q.Set("auth-id", p.authID)
+	q.Set("auth-password", p.authPassword)
+
+	reqURL.RawQuery = q.Encode()
+
+	req, err := http.NewRequest(method, reqURL.String(), http.NoBody)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := p.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("invalid status code %v", resp.Status)
+	}
+
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return body, nil
+}
+
+func (p *cloudnsProvider) GetTimeout() (timeout, interval time.Duration) {
+	return p.dnsPropagationTimeout, p.dnsPropagationInterval
+}
+
+func getEnvOptionalInt(key string, defaultValue int) int {
```

Comment:
those two are likely universal helpers for all providers and probably better to be moved out of the particular implementation, maybe to provider.go ?

Btw, I would strongly suggest adding the second provider. This way we may see issues with the common code and, maybe, with abstractions. My vote is for route53 (seems to be easy enough to implement), but up to you

Reply:
It makes sense. I'll move them to provider.go

I am going to implement route53 provider.  If I did everything correctly with abstraction then It should be a very straightforward task.